### ISSUE TITLE
Python - High num_leaves leads to program crashing even when low value of max_depth is provided.

### ISSUE THREAD

**Issue by dineshdharme** [microsoft/LightGBM#2215]
## Environment info

Operating System: Ubuntu 18.04.2 LTS
CPU/GPU model: Intel(R) Core(TM) i7-7820HK CPU @ 2.90GHz
C++/Python/R version: Python
LightGBM version or commit hash: 2.2.3

## Error message
Process finished with exit code 137 (interrupted by signal 9: SIGKILL)

## Reproducible examples
The following `num_leaves` value leads to OOM and the program gets killed.
If I reduce the `num_leaves` to 20000 then it works. Shouldn't the `max_depth` parameter be respected before the `num_leaves` ? By providing a large `num_leaves` values, I wanted to compare how lightgbm random forest compares with other implementations for the same `max_depth`  

	# specify your configurations as a dict
	params = {
	    'boosting_type': 'rf',
	    'objective': 'binary',
	    'metric': 'binary_logloss',
	    'num_leaves': 2000000,
	    'max_depth':1,
	    'learning_rate': 0.05,
	    'feature_fraction': 0.9,
	    'bagging_fraction': 0.8,
	    'bagging_freq': 5,
	    'verbose': 0,
	    'num_iterations':10
	}

	print('Starting training...')

	model = lgb.LGBMClassifier(**params)
	model.fit(X_train, y_train)

## Steps to reproduce
I am providing a small example which reproduces the issue. 
https://github.com/dineshdharme/lightgbm-issue1

**Comment by guolinke** [comment]
Thanks for pointing out this.
There is a histogram pool size, which is set by num_leaves. 
Fix in #2216

**Comment by dineshdharme** [comment]
Does this ensure that the tree will be expanded to the full `max_depth` ?
For max_depth = 4, I am getting trees like this.
https://imgur.com/YZf0del

**Comment by guolinke** [comment]
if num_leaves >= 2 ** max_depth, and the split gain is positive, it should be fully binary tree.
